### PR TITLE
azure: add related.entity field to auditlogs and signinlogs default ingest pipeline

### PR DESCRIPTION
**Pull request by orouz**
## Proposed commit message

adds processors to auditlogs and signinlogs ingest pipelines that appends all entity identifiers from the log event to related.entity in order to facilitate pivoting around a piece of data

## Context

this PR is part of the cloud security [CDR epic](https://github.com/elastic/security-team/issues/10152). it populates `related.entity` which is an [upcoming ECS field](https://github.com/elastic/ecs/pull/2360)

## benchmark (100 runs)

1. auditlogs

Metric |    |  
-- | -- | --
Description | Main | append + painless script processor
Commit | https://github.com/elastic/integrations/commit/63854f060182ce150072311d30d50281d9218f04 | 01e0f3c
Average EPS | 16409.48978  | 15400.45392
Stddev | 1273.233006  | 1197.630202
Min | 8196.721311  | 10416.66667
Max | 17857.14286 | 16949.15254
EPS Change to baseline |    | -6.15% 

2. signinlogs

Metric |    |  
-- | -- | --
Description | Main | append processor
Commit |https://github.com/elastic/integrations/commit/63854f060182ce150072311d30d50281d9218f04 | de9827c
Average EPS | 9641.707196  | 9561.806718
Stddev | 701.7646604  | 861.931866
Min | 7142.857143  | 5405.405405
Max | 11363.63636 | 10989.01099
EPS Change to baseline |    | -0.83%

## Related issues
- closes https://github.com/elastic/security-team/issues/10391

### CODE REVIEW

File: packages/azure/changelog.yml [hunk 1]
```diff
@@ -1,3 +1,8 @@
+- version: "1.18.0"
+  changes:
+    - description: "Add entity identifiers to `related.entity`"
```
Comment (efd6) [inline packages/azure/changelog.yml:3]:
```suggestion
    - description: Add entity identifiers to `related.entity`.
```

File: packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -229,6 +229,35 @@ processors:
...
+      if: ctx.azure?.auditlogs?.properties?.initiated_by?.app?.servicePrincipalId != null && ctx.azure.auditlogs.properties.initiated_by.app.servicePrincipalId != ''
+  - script:
+      description: Appends target resources identifiers to `related.entity`
+      lang: painless
+      source: |
+        ctx.related = ctx.related ?: [:];
+        ctx.related.entity = ctx.related.entity ?: [];
+        if (ctx.azure.auditlogs.properties.target_resources != null) {
+            for (String i : ctx.azure.auditlogs.properties.target_resources.keySet()) {
+                def resource = ctx.azure.auditlogs.properties.target_resources[i];
```
Comment (efd6) [inline packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml:250]:
```suggestion
            for (String k : ctx.azure.auditlogs.properties.target_resources.keySet()) {
                def resource = ctx.azure.auditlogs.properties.target_resources[k];
```
or perhaps
```suggestion
            for (String id : ctx.azure.auditlogs.properties.target_resources.keySet()) {
                def resource = ctx.azure.auditlogs.properties.target_resources[id];
```
if the key is an ID (it's not clear to me that this is the case).

`i` is an unconventional name for a non-integer variable.

File: packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -229,6 +229,35 @@ processors:
...
+                if (resource != null && resource.id != null && resource.id != '' && !ctx.related.entity.contains(resource.id)) {
```
Comment (efd6) [inline packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml:251]:
```suggestion
                if (resource?.id != null && resource.id != '' && !ctx.related.entity.contains(resource.id)) {
```

File: packages/azure/data_stream/auditlogs/fields/fields.yml [hunk 4]
```diff
@@ -191,3 +191,6 @@
     - name: city_name
       type: keyword
       description: City name.
+- name: related.entity
+  description: "All the entity identifiers related to the document. If the document \ncontains multiple entities, identifiers belonging to different entities \nwill be present. Example identifiers include cloud resource IDs, ARNs, \nemail addresses, or hostnames.\n"
```
Comment (efd6) [inline packages/azure/data_stream/auditlogs/fields/fields.yml:195]:
```suggestion
- name: related.entity
  description: >-
    All the entity identifiers related to the document. If the document contains
    multiple entities, identifiers belonging to different entities will be present.
    Example identifiers include cloud resource IDs, ARNs, email addresses, or hostnames.
```

File: packages/azure/data_stream/signinlogs/fields/fields.yml [hunk 5]
```diff
@@ -252,3 +252,6 @@
     - name: city_name
       type: keyword
       description: City name.
+- name: related.entity
+  description: "All the entity identifiers related to the document. If the document \ncontains multiple entities, identifiers belonging to different entities \nwill be present. Example identifiers include cloud resource IDs, ARNs, \nemail addresses, or hostnames.\n"
```
Comment (efd6) [inline packages/azure/data_stream/signinlogs/fields/fields.yml:256]:
```suggestion
- name: related.entity
  description: >-
    All the entity identifiers related to the document. If the document contains
    multiple entities, identifiers belonging to different entities will be present.
    Example identifiers include cloud resource IDs, ARNs, email addresses, or hostnames.
```

**Review by efd6 — COMMENTED** [review]
@orouz Please revise the proposed commit message to be a stand-alone text that explains what is being done and why. It should preferably not refer to internal issues since it will be read by people who may not have access to those documents. Also note that git commit messages are not markdown, so do not use markdown features.

File: packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (orouz) [reply packages/azure/data_stream/auditlogs/elasticsearch/ingest_pipeline/default.yml:250]: indeed. thanks!

**Review by efd6 — APPROVED** [review]

**Review by zmoog — APPROVED** [review]

**Review by kaiyan-sheng — APPROVED** [review]

**Review by muthu-mps — APPROVED** [review]

**Comment by nicpenning** [comment]
`related.entity` doesn't appear to be a field that will be in compliance with ECS.

https://www.elastic.co/guide/en/ecs/current/ecs-related.html

I have an integration going through the works right now that would like to use something like `related.file.name` but was directed to creating an RFC for adopting that field.

Will this cause issues down the road or those that use the Data Quality dashboard to reveal actual ECS fields? Is there already a precedent set here that any field can live beneath `related.`? Thoughts?

**Comment by orouz** [comment]
> `related.entity` doesn't appear to be a field that will be in compliance with ECS.

it's currently being worked on at https://github.com/elastic/ecs/pull/2360

> Will this cause issues down the road or those that use the Data Quality dashboard to reveal actual ECS fields? Is there already a precedent set here that any field can live beneath `related.`? Thoughts?

@tinnytintin10 can comment more on that.

**Pull request merged**